### ISSUE TITLE
Rust default stack size varies across compilation

### ISSUE THREAD

**Issue by biabbas** [rust-lang/rust#126897]
While working on this issue #122357, I found that setting the RUST_MIN_STACK variable size to 7mb or more, the build is successful. But the default stack size in rust is 8mb. Why do we have to set RUST_MIN_STACK again for this to work?

Steps to create the example directory are in this link https://github.com/rust-lang/rust/issues/122357#issue-2180549643
To verify this please follow these steps.
1. setup the example from above link.
2. Run `cargo build` in the example, this time the build would fail.
3. Export `RUST_MIN_STACK=$((7*1024*1024)` and run `cargo build` again.

**Comment by ChrisDenton** [comment]
The default stack size of new threads is 2mb. However, the default stack size of the main thread varies depending on the platform and possibly platform-dependent build configuration.

**Comment by workingjubilee** [comment]
Rust does not override every last platform-specific behavior. To do so would likely be quite harmful.

**Comment by workingjubilee** [comment]
Actually, wait. I'm confused. For rustc, this shouldn't matter because of the [DEFAULT_STACK_SIZE](https://github.com/rust-lang/rust/blob/d371d17496f2ce3a56da76aa083f4ef157572c20/compiler/rustc_interface/src/util.rs#L52).

**Comment by workingjubilee** [comment]
@biabbas What version of rustc did you test with?

**Comment by workingjubilee** [comment]
No repro on `x86_64-unknown-linux-gnu`.

**Comment by workingjubilee** [comment]
Wait, I misapplied some directions.

It... does repro? What.

**Comment by biabbas** [comment]
> @biabbas What version of rustc did you test with?

I used rust 1.78 and 1.79 stable

**Comment by biabbas** [comment]
> The [default stack size of new threads is 2 MiB](https://doc.rust-lang.org/std/thread/#stack-size). However, the default stack size of the main thread varies depending on the platform and possibly platform-dependent build configuration.

You are right, default stack size for new threads is 2mb. I went through the std thread code and I see that min_stack uses this RUST_MIN_STACK variable to influence the size of thread builder stack size. In this example what stack is overflowed and why RUST_MIN_STACK workaround works for it?

**Comment by biabbas** [comment]
> Actually, wait. I'm confused. For rustc, this shouldn't matter because of the [DEFAULT_STACK_SIZE](https://github.com/rust-lang/rust/blob/d371d17496f2ce3a56da76aa083f4ef157572c20/compiler/rustc_interface/src/util.rs#L52) (not to be confused with the default stack size).

Could you please elaborate and differentiate between the two.

**Comment by biabbas** [comment]
Can rustc be using parallelism( sub threads with limit around 2mb) for program compilation? In that case the error message should emit the actual stack that is overflowed. Or the thread builder stack size is overwriting the main threads stack size.

**Comment by workingjubilee** [comment]
Yes, we deliberately set the stack size by spawning a thread.

However, I would expect we are setting it to 8MB.

We emit a suggestion that suggests setting `RUST_MIN_STACK` to double the previous value.
